go/pkg/rexec/rexec.go

```diff
@@ -332,35 +341,31 @@ func (ec *Context) ExecuteRemotely() {
 	ec.Metadata.RealBytesUploaded = bytesMoved
 	log.V(1).Infof("%s %s> Executing remotely...\n%s", cmdID, executionID, strings.Join(ec.cmd.Args, " "))
 	ec.Metadata.EventTimes[command.EventExecuteRemotely] = &command.TimeInterval{From: time.Now()}
-	eg, ctx := errgroup.WithContext(ec.ctx)
-	var streamOut, streamErr sync.Once
-	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ctx, &repb.ExecuteRequest{
+	// These variables are owned by the progress callback (which is async but not conrrent) until the execution returns.
+	var streaming bool
+	var nOutStreamed, nErrStreamed int64
+	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ec.ctx, &repb.ExecuteRequest{
 		InstanceName:    ec.client.GrpcClient.InstanceName,
 		SkipCacheLookup: !ec.opt.AcceptCached || ec.opt.DoNotCache,
 		ActionDigest:    ec.Metadata.ActionDigest.ToProto(),
 	}, func(md *repb.ExecuteOperationMetadata) {
-		if !ec.opt.StreamOutErr {
+		if streaming || !ec.opt.StreamOutErr {
 			return
 		}
+		streaming = true
 		if name := md.GetStdoutStreamName(); name != "" {
-			streamOut.Do(func() {
-				eg.Go(func() error {
-					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
-					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
-					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewOutWriter(ec.oe))
-					return err
-				})
-			})
+			path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
+			log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
+			// Ignoring the error here since the net result is downloading the full stream after the fact.
+			n, _ := ec.client.GrpcClient.ReadResourceTo(ec.ctx, path, outerr.NewOutWriter(ec.oe))
```

Comment:
This will try to download the whole stdout stream synchronously while the first callback is being invoked (and then synchronously download the entire stderr stream below). I don't think this is what we want here. We still need to start a goroutine for each stream.

Note that this callback function will typically be called multiple times during execution to update the caller on the progress of the action. We don't want to hold up the first such call trying to download the whole of both stream synchronously, one after the other.

Since we don't care about handling the errors below (we only need to know how much was successfully streamed), perhaps use a sync.WaitGroup instead of an errgroup?

Reply:
Good catch! I was preoccupied with the fact that the callback is not concurrent that I forgot this call is synchronous.